File: tcg-java/test/de/kimminich/kata/tcg/matchers/MoveMatchers.java
package de.kimminich.kata.tcg.matchers;

import de.kimminich.kata.tcg.Action;
import de.kimminich.kata.tcg.Card;
import de.kimminich.kata.tcg.Move;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.Optional;

import static de.kimminich.kata.tcg.Action.DAMAGE;
import static de.kimminich.kata.tcg.Action.HEALING;

public class MoveMatchers {

    public static Matcher<Move> isAttackingWithCard(int card) {
        return isPerformingActionWithCard(card, DAMAGE);
    }

    public static Matcher<Move> isHealingWithCard(int card) {
        return isPerformingActionWithCard(card, HEALING);
    }

    public static Matcher<Move> isPerformingActionWithCard(int card, Action action) {
        return new TypeSafeMatcher<Move>() {
            @Override
            public boolean matchesSafely(Move move) {
                Optional<Card> moveCard = move.getCard();
                return moveCard.isPresent() && moveCard.get().getManaCost() == card && move.getAction().equals(action);
            }

            @Override
            public void describeTo(Description description) {
                description.appendValue(action).appendText(" with card ").appendValue(card);
            }

            @Override
            public void describeMismatchSafely(Move move, Description description) {
                description.appendText("was ").appendValue(move.getAction())
                        .appendText(" with card ").appendValue(move.getCard().get());
            }
        };
    }

}

File: tcg-kotlin/src/test/kotlin/de/kimminich/kata/tcg/matchers/MoveMatchers.kt
package de.kimminich.kata.tcg.matchers

import de.kimminich.kata.tcg.Action
import de.kimminich.kata.tcg.Move
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher

object MoveMatchers {
    fun isAttackingWithCard(card: Int): Matcher<Move> {
        return isPerformingActionWithCard(card, Action.DAMAGE)
    }

    fun isHealingWithCard(card: Int): Matcher<Move> {
        return isPerformingActionWithCard(card, Action.HEALING)
    }

    private fun isPerformingActionWithCard(card: Int, action: Action): Matcher<Move> {
        return object : TypeSafeMatcher<Move>() {
            public override fun matchesSafely(move: Move): Boolean {
                val moveCard = move.card
                return (moveCard.value == card) && (move.action == action)
            }

            override fun describeTo(description: Description) {
                description.appendValue(action).appendText(" with card ").appendValue(card)
            }

            public override fun describeMismatchSafely(move: Move, description: Description) {
                description.appendText("was ").appendValue(move.action)
                    .appendText(" with card ").appendValue(move.card)
            }
        }
    }
}
